pkg/validation/admission_test.go

```diff
@@ -0,0 +1,139 @@
+	// Register our package types with the global scheme
+	metal3apis.AddToScheme(scheme.Scheme)
+}
+
+func newHost(name, bmcAddress, bmcCredentialsName, bootMACAddress string) metal3v1alpha1.BareMetalHost {
+	return metal3v1alpha1.BareMetalHost{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "BareMetalHost",
+			APIVersion: "metal3.io/v1alpha1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		Spec: metal3v1alpha1.BareMetalHostSpec{
+			BMC: metal3v1alpha1.BMCDetails{
+				Address:         bmcAddress,
+				CredentialsName: bmcCredentialsName,
+			},
+			BootMACAddress: bootMACAddress,
+		},
+	}
+}
+
+func newDefaultHost() metal3v1alpha1.BareMetalHost {
+	return newHost(defaultHostName, defaultBMCAddress, defaultBMCCredentialsName, defaultBootMACAddress)
+}
+
+func TestCanBeAdmitted(t *testing.T) {
+	testCases := []struct {
+		Scenario string
+		Host     metal3v1alpha1.BareMetalHost
+		Expected error
+	}{
+		{
+			Scenario: "BMC address is not unique",
+			Host:     newHost(defaultHostName, defaultBMCAddress, "", ""),
+			Expected: apierrors.NewInvalid(
+				metal3v1alpha1.SchemeGroupVersion.WithKind("BareMetalHost").GroupKind(),
+				defaultHostName,
+				field.ErrorList{
+					field.Invalid(
+						field.NewPath("spec", "bmc", "address"),
+						defaultBMCAddress,
+						"is not unique",
+					),
+				}),
+		},
+		{
+			Scenario: "BMC address is unique",
+			Host:     newHost(defaultHostName, "unique", "", ""),
+			Expected: nil,
+		},
+		{
+			Scenario: "BootMACAddress is not unique",
+			Host:     newHost(defaultHostName, "", "", defaultBootMACAddress),
+			Expected: apierrors.NewInvalid(
+				metal3v1alpha1.SchemeGroupVersion.WithKind("BareMetalHost").GroupKind(),
+				defaultHostName,
+				field.ErrorList{
+					field.Invalid(
+						field.NewPath("spec", "bootMACAddress"),
+						defaultBootMACAddress,
+						"is not unique",
+					),
+				}),
+		},
+		{
+			Scenario: "BootMACAddress is unique",
+			Host:     newHost(defaultHostName, "", "", "unique"),
+			Expected: nil,
+		},
+		{
+			Scenario: "BMC credentials name is not unique",
+			Host:     newHost(defaultHostName, "", defaultBMCCredentialsName, ""),
+			Expected: apierrors.NewInvalid(
+				metal3v1alpha1.SchemeGroupVersion.WithKind("BareMetalHost").GroupKind(),
+				defaultHostName,
+				field.ErrorList{
+					field.Invalid(
+						field.NewPath("spec", "bmc", "credentialsName"),
+						defaultBMCCredentialsName,
+						"is not unique",
+					),
+				}),
+		},
+		{
+			Scenario: "BMC credentials name is unique",
+			Host:     newHost(defaultHostName, "", "unique", ""),
+			Expected: nil,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.Scenario, func(t *testing.T) {
+			existingHost := newDefaultHost()
+			c := fakeclient.NewFakeClient(&existingHost)
+			err := ValidateHostForAdmission(c, tc.Host)
+			assert.Equal(t, tc.Expected, err)
+		})
+	}
+}
+
+func TestCanBeAdmittedMultiple(t *testing.T) {
```

Comment:
This should also test admission for a host with no values set. Passing them to the fake client won't do that, right?

Reply:
Done